### DISCUSSION TITLE
Handling Firebase Dynamic Links with expo-router

### DISCUSSION THREAD

**Discussion by naporin0624** [expo/router#656] [Q&A]
I'm using Firebase Dynamic Links in my project.

With the universal link feature of `expo-router`, when a URL generated by Firebase Dynamic Links is `https://xxxxxxx.page.link/details`, it opens `app/details.tsx`.

I would like to know how to open `app/index.tsx` instead in these situations.

Furthermore, is it possible to disable the universal link feature of `expo-router`?
By "disabling", I mean that the app always navigates to `app/index.tsx`.

**Comment by avk-holion** [comment]
I have the same problem, i would like to disable the auto routing on universal link, but there seems to be no way to describe linking options. Please let me know if you find a soloution.

**Comment by GibbyBox** [comment]
I too also need this and neither of the listeners on their docs gets called. [Listening for Dynamic Links](https://rnfirebase.io/dynamic-links/usage#listening-for-dynamic-links)

I also tried using `dynamicLinks().resolveLink()` to try to force a route change but firebase kept throwing `dynamicLinks/not-found`. I'm sure I set up firebase correctly with their plugin. This was working before I swapped to expo-router.

```
	useEffect(function handleDynamicLinks() {
		Linking.addEventListener('url', async ({ url }) => {
			console.debug('handleDynamicLinks', url);

			try {
				const dynamicLink = await dynamicLinks().resolveLink(url);
				router.push(dynamicLink.url);
			} catch (err) {
				console.error(err);
			}
		});
	}, []);
```

**Comment by aledebla03** [comment]
no solution?

**Comment by GibbyBox** [comment]
I'm aware this isn't going to be stuff developers landing on this want to hear, but be aware that google announced they will shut down dynamic links in the future.
https://firebase.google.com/support/dynamic-links-faq

It's still a ways away that I think this is still something to look at for people using dynamic links heavily, but if **you're just starting to use dynamic links**, I would recommend swapping your deferred deep linking needs to another provider.

**Comment by aledebla03** [comment]
is there a way to disable automatic deep link?

**Reply by angelopedroso** [reply]
Did u find a solution for it?

**Reply by GibbyBox** [reply]
I haven't played around with this, but this _should_ handle the issue.
https://docs.expo.dev/router/advanced/native-intent/#rewrite-incoming-native-deep-links

**Comment by LunatiqueCoder** [comment]
🔗 **SOURCE:** https://docs.expo.dev/router/advanced/native-intent/#rewrite-incoming-native-deep-links


<img width="600" alt="image" src="https://github.com/user-attachments/assets/f3ef56fc-7021-4b0d-93a3-6c6b4826c1bb" />

## Solution: 

### 1. Paste this code in `+native-intent.tsx`. Adapt it if necessary.

```js
export function redirectSystemPath({ path }) {
  try {
    // Handle dynamic links
    if (path.includes('.page.link/')) {
     // Handle dynamic links on index
      return `/?${new URLSearchParams({ dynamicLink: path }).toString()`;
    }

    return path;
  } catch {
    // Do not crash inside this function! Instead you should redirect users
    // to a custom route to handle unexpected errors, where they are able to report the incident
    return '/unexpected-error';
  }
}
```

### 2. Then, in your `index.ts` screen:

```ts
  const { dynamicLink } = useLocalSearchParams<{ dynamicLink: string }>();

  useEffect(() => {
    const handleLink = (link: FirebaseDynamicLinksTypes.DynamicLink) => {
     // ... your logic based on the dynamic link
    };

    if (dynamicLink) {
      dynamicLinks()
        .resolveLink(dynamicLink)
        .then(handleLink)
    } 
  }, [dynamicLink]);

```
